### ISSUE TITLE
gatsby-transformer-leasot: Allow other file types than "File"

### ISSUE THREAD

**Issue by muuvmuuv** [GaiAma/Coding4GaiAma#21]
I have created a plugin to parse directories only once which is adapted from `gatsby-plugin-filesystem`. Unofrtunatelly you have explicity set the `source.internal.type` to file, so I have no chance to pass them to `gatsby-transformer-leasot`.

I expect something like you did with `sourceInstanceName`, e.g.: `internalType`

**Comment by CanRau** [comment]
Interesting, uh now I get it^^
Sure that sounds legit 👍
I'll look into it

**Comment by CanRau** [comment]
We're just talking about turning this
```js
node.internal.type === 'File'
```
into this
```js
node.internal.type === internalType
```
right?

Then you set the `sourceInstanceName` to the `node.sourceInstanceName` of your nodes and the `internalType` to your type correct?

Thinking about documentation for this.

Please let me know if that's what you mean/need, I'll finish docs in the meantime and release once you confirm 😉

**Comment by muuvmuuv** [comment]
Yes, exactly! :) That is the only way to get custom source types to work, since  gatsby does not allow the same type from two plugins.

**Comment by CanRau** [comment]
Released as v1.2.0 🎉 

By the way I like your future visions from your [about page](https://marvin.digital/about) 👏

**Comment by CanRau** [comment]
v1.2.2 updates dependencies and switches from "prepublish" to "prepublishOnly" which should fix a possible run of the build step after installing.

**Comment by muuvmuuv** [comment]
Thank you! :) And yes, this works! Btw. the GitHub release version and NPM version mismatch and I think you haven't commited yet. I was a little confused when I searched for the updated part here :D.

I noticed that the `File` prop is always empty. Any idea why this happens?

![Bildschirmfoto 2020-03-01 um 11 02 56](https://user-images.githubusercontent.com/11534760/75623642-477b8000-5bac-11ea-8720-af370fee1f24.png)

If it isn't a quick fix, I will open a new issue.

**Comment by CanRau** [comment]
Yea haven't pushed yet 😅sorry for the confusion

Could you do me a favor and change line 138 of node_modules/gatsby-transformer-leasot/gatsby-node.js from
```graphql
file: File @link(by: "id", from: "file___NODE")
```
to
```graphql
file: NameOfYourInternalType @link(by: "id", from: "file___NODE")
```
`NameOfYourInternalType` just as a string to test locally if that works, hm could maybe even be the `file___NODE` part.
I'm pretty sure it's this line failing because of your custom `node.internal.type`.

Not sure as I haven't linked anything other than File so far & don't know your source plugin.

Let me know how that goes than I should be able to make it dynamic based on `internalType`

Also you could try to get to the parent via an inline fragment like so
```graphql
file {
  ... on File {
    relativePath
    changeTime
  }
}
```
Where you replace `on File` by your `node.internal.type`
I've written about that here https://coding4.gaiama.org/en/graphql-parent-file
Not sure it works with that 'broken' key tho. Not a GraphQL expert 😅

**Comment by muuvmuuv** [comment]
Hey, sorry for the late response! Yes, it is `NameOfYourInternalType`. I changed it to:

```js
exports.createSchemaCustomization = ({ actions }, { sourceInstanceName = defaultSourceInstanceName, internalType = defaultInternalType }) => {
    const { createTypes } = actions;
    const capitalizedInstanceName = sourceInstanceName[0].toUpperCase() + sourceInstanceName.slice(1);
    const typeDefs = `
    type ${capitalizedInstanceName} implements Node @derivedTypes @dontInfer {
      todo: ${capitalizedInstanceName}Todo
    }

    type ${capitalizedInstanceName}Todo {
      tag: String
      line: Int
      ref: String
      text: String
      file: ${internalType} @link(by: "id", from: "file___NODE")
      value: String
      modifiedTime: Date @dateformat
    }
  `;
    createTypes(typeDefs);
};
```

**Comment by CanRau** [comment]
So this works for you correct?

**Comment by muuvmuuv** [comment]
Yes, this works.

**Comment by CanRau** [comment]
Published as **v1.3.0** 🎉 

Thank you Marvin 🙏

**Comment by muuvmuuv** [comment]
Thank you for the work! You can see the final result here: https://marvin.digital/issues (haven't updated the links/refs yet)

**Comment by CanRau** [comment]
You're welcome 😃 thanks for the link 🙏
Looks great 👏
And I love the credits page, have to do that as well ✊
